src/core/js/tracking.js

```diff
@@ -0,0 +1,111 @@
+define([
+    'core/js/adapt',
+    'core/js/enums/completionStateEnum'
+], function(Adapt, COMPLETION_STATE) {
+
+    var Tracking = Backbone.Controller.extend({
+
+        _config: {
+            _requireContentCompleted: true,
+            _requireAssessmentCompleted: false
+        },
+
+        _assessmentState: null,
+
+        initialize: function() {
+            Adapt.once('configModel:loadCourseData', this.loadConfig.bind(this));
+            Adapt.on('app:dataReady', this.setupEventListeners.bind(this));
+        },
+
+        setupEventListeners: function() {
+            // Check if completion requires passing an assessment. 
+            if (this._config._requireAssessmentCompleted) {
+                this.listenTo(Adapt, 'assessment:complete', this.onAssessmentComplete);
+            }
+
+            // Check if completion requires completing all content.
+            if (this._config._requireContentCompleted) {
+                this.listenTo(Adapt.course, 'change:_isComplete', this.checkCompletion);
+            }
+        },
+
+        /**
+         * Store the assessment state.
+         * @param {object} assessmentState - The object returned by Adapt.assessment.getState()
+         */
+        onAssessmentComplete: function(assessmentState) {
+            this._assessmentState = assessmentState;
+
+            this.checkCompletion();
+        },
+
+        /**
+         * Evaluate the course and assessment completion.
+         */
+        checkCompletion: function() {
+            var completionData = this.getCompletionData();
+
+            if (completionData.status === COMPLETION_STATE.INCOMPLETE) {
+                return;
+            }
+
+            Adapt.trigger('tracking:complete', completionData);
+            Adapt.log.debug('tracking:complete', completionData);
+        },
+
+        /**
+         * The return value of this function should be passed to the trigger of 'tracking:complete'.
+         * @returns An object representing the user's course completion.
+         */
+        getCompletionData: function() {
+            var completionData = {
+                status: COMPLETION_STATE.INCOMPLETE,
+                assessment: null
+            };
+
+            // Course complete is required.
+            if (this._config._requireCourseComplete && !Adapt.course.get('_isComplete')) {
+                // INCOMPLETE: course not complete.
+                return completionData;
+            }
+
+            // Assessment completed required.
+            if (this._config._requireAssessmentCompleted) {
+                if (!this._assessmentState) {
+                    // INCOMPLETE: assessment is not complete.
+                    return completionData;
+                }
+
+                if (!this._assessmentState.isPass && this._assessmentState.attempts) {
```

Comment:
`_assessmentState` doesn't have an `attempts` property, which means that if you have the criteria set to `_requireAssessmentCompleted` only and the assessment configured to have infinite attempts, you get set to a FAILED status as soon as you fail.

Reply:
appreciate this is more an issue with the assessment extension that with this code...

Reply:
There should be an `attempts` property returned by `getState()` here:
https://github.com/adaptlearning/adapt-contrib-assessment/blob/master/js/adapt-assessmentArticleModel.js#L777

In this case there's an early return, with a `status` of incomplete set.  The FAILED status is only set after a fail on the last attempt.

Reply:
I think that's the state you get when _an_ assessment is completed not when _all_ of them are completed.
This is the `assessmentState` object that `onAssessmentComplete` in tracking.js gets:
![pr_1797_assessment_state](https://user-images.githubusercontent.com/1249597/32678384-05f9da3a-c65a-11e7-9566-e89bbe921d3a.png)
To be honest, we're actually no worse off here than we are with the current Adapt code; it does exactly the same thing. Really the only issue here is that the code is trying to check something that doesn't actually exist, the outcome is no different to what we currently have.